lib/passport-azure-ad/oidcstrategy.js

```diff
@@ -0,0 +1,203 @@
+
+  if (options.issuer) {
+    log.info('Issuer provided to Strategy was: ', options.issuer);
+  }
+
+  // if you want to check JWT audience (aud), provide a value here
+
+  if (options.audience) {
+    log.info('Audience provided to Strategy was: ', options.audience);
+  }
+
+  if (options.identityMetadata) {
+
+    log.info('Metadata url provided to Strategy was: ', options.identityMetadata);
+    this.metadata = new Metadata(options.identityMetadata, "oidc");
+  }
+
+  if (!options.certificate && !options.identityMetadata) {
+    log.warn("No options was presented to Strategy as required.");
+    throw new TypeError('OIDCBearerStrategy requires either a PEM encoded public key or a metadata location that contains cert data for RSA and ECDSA callback.');
+  }
+
+  if (typeof options === 'function') {
+    verify = options;
+    options = {};
+  }
+
+  // Passport requires a verify function
+
+  if (!verify) {
+    throw new TypeError('OIDCBearerStrategy requires a verify callback. Do not cheat!');
+  }
+
+  // Token validation settings. Hopefully most of these will be pulled from the metadata and this is not needed
+
+
+  this.metadata.fetch(function(err) {
+    if (err) {
+      throw new Error("Unable to fetch metadata: " + err);
+    }
+
+  });
+
+
+  function jwtVerify(req, token, done) {
+
+    if (!options.passReqToCallback) {
+      token = arguments[0];
+      done = arguments[1];
+      req = null;
+      log.info(token, 'was token going in to verification');
+    }
+
+
+    var decoded = jws.decode(token);
+    if (decoded == null) {
+      done(null, false, "Invalid JWT token.");
+    }
+
+    log.info('token decoded. But is it valid?: ', decoded);
+
+
+    // We have two different types of token signatures we have to validate here. One provides x5t and the other a kid.
+    // We need to call the right one.
+
+    if (decoded.header.x5t) {
+      PEMkey = this.metadata.generateOidcPEM(decoded.header.x5t);
+    } else if (decoded.header.kid) {
+      PEMkey = this.metadata.generateOidcPEM(decoded.header.kid);
+    } else {
+      throw new TypeError('We did not reveive a token we know how to validate');
+    }
+
+
+
+    if (!options.issuer) {
+      options.issuer = this.metadata.oidc.issuer;
+    }
+    options.algorithms = this.metadata.oidc.algorithms;
+
+    jwt.verify(token, PEMkey, options, function(err, token) {
+      if (err) {
+        if (err instanceof jwt.TokenExpiredError) {
+          log.warn("Access token expired");
+          done(null, false, 'The access token expired');
+        } else if (err instanceof jwt.JsonWebTokenError) {
+          log.warn("An error was received validating the token", err.message);
+          done(null, false, util.format('Invalid token (%s)', err.message));
+        } else {
+          done(err, false);
+        }
+      } else {
+        log.info(token, 'was token going out of verification');
```

Comment:
I can remove the token logging if that's the standard. I figured it was helpful for first time users and can be removed by the developer. No?